Add image-check-module-sigs.bbclass

jandryuk opened the pull request:
image-check-module-sigs.bbclass checks the signatures of modules
installed in an image to ensure they are valid for the associated
kernel.  Usually this is /boot/bzImage, but it will check in deploy dir
for stubdomain's external kernel.

This adds ~7 seconds to dom0 image creation, but it's nicer to check during
the build than have an install that fails to boot properly.

Tie it in to openxt-image so the check is always performed.

Signed-off-by: Jason Andryuk <jandryuk@gmail.com>

The installer doesn't have signed modules, so there is a workaround (if [[ $output =~ "Can't find system certificate list " ]]; then) to skip the check in that case.

pipefail is enabled so we get the exit value of `openssl smime -verify` on failure and not the (always successful) exit status of `sed -n '/Verification successful/!p'`.  I did that so the log will have any error message from smime without spamming the log with Verification successful.  The caller will print the failing module name, so silencing all output from smime is another option.  The error output from smime isn't terribly informative when a signature fails, but there could be value there.

jandryuk commented on classes/image-check-module-sigs.bbclass:
```diff
@@ -0,0 +1,89 @@
+	if ! output=$( $scriptdir/extract-sys-certs.pl \
+			-s $builddir/System.map-$vers $vmlinux \
+			$keyring 2>&1 1>/dev/null ) ; then
+		if [[ $output =~ "Can't find system certificate list " ]]; then
```
Yes, it's a bash-ism.  I can change to `expr "$output" : "^Can't find system certificate list "`

jandryuk commented on classes/image-check-module-sigs.bbclass:
```diff
@@ -0,0 +1,89 @@
+do_image_check_module_sigs() {
+	bzImage=${IMAGE_ROOTFS}/boot/bzImage
```
Will do.

jandryuk commented:
I noticed this morning that image.bbclass includes task do_image_qa which executes the commands in IMAGE_QA_COMMANDS.  Should I change the explicit do_image_check_module_sigs task to just add to IMAGE_QA_COMMANDS?  Maybe s/do_image_check_module_sigs/do_image_qa_module_sigs/ as well?

eric-ch commented:
@jandryuk `IMAGE_QA_COMMANDS` looks a lot more streamlined, I did not know about it, +1 on the renaming of the function as well.

Looking at the history, it could be leveraged more to have builds fail earlier (e.g, checking that `tmpfs` directory mount points are not populated by packages).

jandryuk commented:
Updated with:

- Quoted variables
- Rename to image_qa_module_sigs
- Replace `[[` bash-ism with `expr`
- Use do_image_qa/IMAGE_QA_COMMANDS
- Added do_image_qa[depends] to ensure we have perl for the linux extract scripts and openssl which is called directly.

jandryuk commented:
Also, I pass keyring as a parameter to check_module instead of relying on it being set globally.

eric-ch commented on classes/image-qa-module-sigs.bbclass:
```diff
@@ -0,0 +1,95 @@
+	return "$ret"
+}
+
+do_image_qa[depends] += "perl-native:do_populate_sysroot openssl-native:do_populate_sysroot"
```
`$scriptdir` will be provided through the `$tmpdir/work-shared` so this should add `virtual/kernel:do_shared_workdir` to `depends`.

jandryuk replied:
Yes, it should - both for the scripts and System.map-$vers.  I am concerned that if SSTATE is used, bzImage comes from the linux-openxt package, but System.map will be from a newly compiled copy.  They *should* be the same, but it makes me uncomfortable.  System.map is packaged in kernel-dev with:
boot/Module.symvers-5.10.25
boot/System.map-5.10.25
boot/config-5.10.25

Maybe instead of installing the package, we could do_deploy System.map?

eric-ch commented on classes/image-qa-module-sigs.bbclass:
```diff
@@ -0,0 +1,95 @@
+do_image_qa_module_sigs() {
+	bzImage="${IMAGE_ROOTFS}/boot/bzImage"
```
I missed it during the first review, variables could be made `local`, yet it does not look like this is idiomatic in OE recipes.

jandryuk commented:
Updated `do_image_qa[depends]` and made variables local.

(Had a bad force push in there on top of vglass and usbvm).

I'm still unsure what to do about System.map.

eric-ch reviewed (approved):
I have been using this a fair amount while testing https://github.com/OpenXT/xenclient-oe/pull/1401.
LGTM.
One potential caveat: (https://github.com/OpenXT/xenclient-oe/pull/1401#discussion_r622414056), although I do not want to block this on such a corner case unless someone has a current use for it.

eric-ch commented:
Will merge soon.
